src/Ouroboros/Network/Pipe.hs

```diff
@@ -207,7 +211,11 @@ runProtocolWithPipe hndRead hndWrite p =
           -- print ("Recv", msg)
           k msg >>= go trailing'
 
-    go _trailing (Fail failure) = fail (show failure)
+    go _trailing (Fail failure) = do
```

Comment:
Printing the failure and returning unit is not very satisfactory. Let's throw an exception instead of using `fail`, and then we can just catch it when needed.

Reply:
`throwIO failure`, will need to make the failure type an instance of `Exception`.